src/network/network_gui.cpp

```diff
@@ -1806,166 +1793,543 @@ static void PopupClientList(ClientID client_id, int x, int y)
 		}
+
+		return x;
 	}
 
-	void OnClick(Point pt, int widget, int click_count) override
+	uint DrawCompany(const Company *c, uint left, uint right, uint top, uint count) const
 	{
-		/* Show the popup with option */
-		if (this->selected_item != -1) {
-			int client_no = this->selected_item;
-			for (NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-				if (client_no == 0) {
-					PopupClientList(ci->client_id, pt.x + this->left, pt.y + this->top);
-					break;
+		bool rtl = _current_text_dir == TD_RTL;
+		uint text_y_offset = (this->line_height - FONT_HEIGHT_NORMAL) / 2;
+
+		Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
+		uint text_left = left + (rtl ? (uint)WD_FRAMERECT_LEFT : d.width + 8);
+		uint text_right = right - (rtl ? d.width + 8 : (uint)WD_FRAMERECT_RIGHT);
+
+		uint count_start = this->vscroll->GetPosition();
+		uint count_end = count_start + this->vscroll->GetCapacity();
+
+		uint y = top + (this->line_height * (count - count_start));
+
+		/* Draw the company line (if in range of scrollbar). */
+		if (count >= count_start && count < count_end) {
+			uint x = rtl ? text_left : text_right;
+
+			/* If there are buttons for this company, draw them. */
+			auto button_find = this->buttons.find(count);
+			if (button_find != this->buttons.end()) {
+				x = this->DrawButtons(x, y, &button_find->second);
+			}
+
+			if (c == nullptr) {
+				DrawString(rtl ? x : text_left, rtl ? text_right : x, y + text_y_offset, STR_NETWORK_CLIENT_LIST_SPECTATORS, TC_BLACK);
+			} else {
+				DrawCompanyIcon(c->index, rtl ? right - d.width - 2 : left + 2, y + (this->line_height - d.height) / 2);
+
+				SetDParam(0, c->index);
+				SetDParam(1, c->index);
+				DrawString(rtl ? x : text_left, rtl ? text_right : x, y + text_y_offset, STR_COMPANY_NAME_COMPANY_NUM, TC_BLACK);
+			}
+		}
+
+		y += this->line_height;
+		count++;
+
+		bool has_players = false;
+		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
+			if (c != nullptr && ci->client_playas != c->index) continue;
+			if (c == nullptr && ci->client_playas != COMPANY_SPECTATOR) continue;
+			if (*ci->client_name == '\0') continue;
+			has_players = true;
+
+			/* Draw the player line (if in range of scrollbar). */
+			if (count >= count_start && count < count_end) {
+				uint x = rtl ? text_left : text_right;
+
+				/* If there are buttons for this client, draw them. */
+				auto button_find = this->buttons.find(count);
+				if (button_find != this->buttons.end()) {
+					x = this->DrawButtons(x, y, &button_find->second);
+				}
+
+				StringID client_string = STR_JUST_RAW_STRING;
+
+				if (ci->client_id == CLIENT_ID_SERVER) {
+					client_string = STR_NETWORK_CLIENT_LIST_PLAYER_HOST;
+				}
+				if (ci->client_id == _network_own_client_id) {
+					client_string = STR_NETWORK_CLIENT_LIST_PLAYER_SELF;
+				}
+
+				SetDParamStr(0, ci->client_name);
+				DrawString(rtl ? x : text_left + 12, rtl ? text_right - 12 : x, y + text_y_offset, client_string, TC_BLACK);
```

Comment:
Do the repeated 12s here and around line 2298 have the "same" function for aligning things, then you probably should make it a constant and use that constant in all the places where that applies.